File: 01-two-number-sum/elixir/solution.exs
defmodule Solution do
  def two_number_sum_solution(numsArr, targetSum) do
    case two_number_sum_reduce(numsArr, targetSum) do
      result when is_list(result) -> result
      _ -> []
    end
  end

  defp two_number_sum_reduce(numsArr, targetSum) do
    Enum.reduce_while(numsArr, %{}, fn num, acc ->
      potentialMatch = targetSum - num

      if Map.get(acc, potentialMatch) do
        {:halt, [potentialMatch, num]}
      else
        {:cont, Map.put(acc, num, true)}
      end
    end)
  end

  def two_number_sum_solution_01(numsArr, targetSum) do
    two_number_sum_solution_01(numsArr, targetSum, %{})
  end

  defp two_number_sum_solution_01([], _targetSum, _numsMap), do: []

  defp two_number_sum_solution_01([num | tail], targetSum, numsMap) do
    potentialMatch = targetSum - num

    case Map.get(numsMap, potentialMatch) do
      nil -> two_number_sum_solution_01(tail, targetSum, Map.put(numsMap, num, true))
      _result -> [potentialMatch, num]
    end
  end
end

arr = [-4, -1, 1, 3, 5, 6, 8, 11]
target = 10
Solution.two_number_sum_solution(arr, target) |> IO.inspect()
Solution.two_number_sum_solution_01(arr, target) |> IO.inspect()

File: 01-two-number-sum/go/solution.go
package main

import (
	"fmt"
	"slices"
)

func main() {
	slice := []int{-4, -1, 1, 3, 5, 6, 8, 11}
	target := 10

	fmt.Println(twoNumberSumSolution(slice, target))
	fmt.Println(twoNumberSumSolution01(slice, target))
	fmt.Println(twoNumberSumSolution02(slice, target))
}

func twoNumberSumSolution(numsSlice []int, targetSum int) []int {
	for i := 0; i < len(numsSlice)-1; i++ {
		firstNum := numsSlice[i]
		for j := i + 1; j < len(numsSlice); j++ {
			secondNum := numsSlice[j]
			if firstNum+secondNum == targetSum {
				return []int{firstNum, secondNum}
			}
		}
	}
	return []int{}
}

func twoNumberSumSolution01(numsSlice []int, targetSum int) []int {
	numsMap := make(map[int]bool)
	for _, num := range numsSlice {
		potentialMatch := targetSum - num
		_, ok := numsMap[potentialMatch]
		if ok {
			return []int{potentialMatch, num}
		} else {
			numsMap[num] = true
		}
	}
	return []int{}
}

func twoNumberSumSolution02(numsSlice []int, targetSum int) []int {
	slices.Sort(numsSlice)
	left := 0
	right := len(numsSlice) - 1
	for left < right {
		currentSum := numsSlice[left] + numsSlice[right]
		if currentSum == targetSum {
			return []int{numsSlice[left], numsSlice[right]}
		} else if currentSum < targetSum {
			left += 1
		} else if currentSum > targetSum {
			right -= 1
		}
	}
	return []int{}
}
